core/src/main/java/org/infinispan/persistence/sifs/Compactor.java

```diff
@@ -480,11 +487,8 @@ private void compactSingleFile(CompactionRequest compactionRequest,
       try (handle) {
          long fileSize = handle.getFileSize();
          AggregateCompletionStage<Void> aggregateCompletionStage = CompletionStages.aggregateCompletionStage();
-         while (!clearSignal.get() && !terminateSignal) {
-            EntryHeader header = EntryRecord.readEntryHeader(handle, scheduledOffset);
-            if (header == null) {
-               break;
-            }
+         EntryHeader header;
+         while ((header = EntryRecord.readEntryHeader(handle, scheduledOffset)) != null) {
```

Comment:
Could this delay a node shutdown? Or do we usually see small files here?

Reply:
Yes, that is the point of the change as we have to completely compact the current file otherwise it will lead to having a corrupt index (note pending compactions will not be processed, only the current). If the user forcibly stops the server then the index will just be dirty and require rebuilding the entire index which is way worse.